### ISSUE TITLE
Isolate WebSocket Implementations

### ISSUE THREAD

**Issue by blackspherefollower** [blackspherefollower/buttplug4j#29]
I don't think the javax provider model is working as I'd like it to when embedded into other applications, and instead it's causing problems that would probably not exist with simpler library structures.

**Comment by AaxAisa** [comment]
This is very much needed. The library doesn't work at all in Android project, for example, because Andreoid's SDK lacks some key things from JDK11 that Jetty is built on. Namely, ServiceLoader.Provider doesn't exist on Android, which maked the entire thing fall apart like a house of cards whenever you try to establish connection to server.
I don't know if there's a workaround - I never poked into websockets and networking at all - it's like black magic to me :D

**Comment by blackspherefollower** [comment]
@AaxAisa Are you up for testing a refactor?
```
repositories {
	maven { url 'https://s01.oss.sonatype.org/content/repositories/snapshots' }
}
dependencies {
    api 'io.github.blackspherefollower:buttplug4j.connectors.jetty.websocket.client:3.1.101-SNAPSHOT'
}
```

**Comment by AaxAisa** [comment]
Sure. I may not have the time today but tomorrow I'll try it for sure!

**Comment by AaxAisa** [comment]
It seems to work!

**Comment by blackspherefollower** [comment]
@AaxAisa did you edit your comment? Just wondering what issue you worked around there

**Comment by AaxAisa** [comment]
Yes, so initially I was getting the same error. But that was because it was working through old version of ButtplugClientWSClient. Gradle kind of installed them separately. When I changed it - it started working.

That is, kind of.
I can connect to the locally running (On Android) Intiface Central. It returns the connection, it even returns the list of devices with the correct device. But something seems to be broken in the message exchange there.

See, it does return the list of devices, but only if I go to Intiface and manually search and connect to device.
And even though it has found the device, no commands actually work on it - i.e. I can't turn it on.

If I do not manually connect to the device, then what it is supposed to do is make Intiface search for devices, find them and report them. That happens for me on Desktop (same code, but desktop app). It does not happen on Android. Instead, if I don't manually connect to a device it reports bad WS.
`io.github.blackspherefollower.buttplug4j.client.ButtplugClientException: Bad WS state!
2023-08-08 21:39:54.952 19169-19169 System.err              com.<packagename>               W  	at io.github.blackspherefollower.buttplug4j.client.ButtplugClient.requestDeviceList(ButtplugClient.java:162)
2023-08-08 21:39:54.952 19169-19169 System.err              com.<packagename>               W  	at com.gir.keyoncommander.MainActivity.startAndRun(MainActivity.java:80)
2023-08-08 21:39:54.952 19169-19169 System.err              com.<packagename>               W  	at com.gir.keyoncommander.MainActivity.access$000(MainActivity.java:19)
2023-08-08 21:39:54.952 19169-19169 System.err              com.<packagename>               W  	at com.gir.keyoncommander.MainActivity$1.onClick(MainActivity.java:39)
2023-08-08 21:39:54.952 19169-19169 System.err              com.<packagename>               W  	at android.view.View.performClick(View.java:7317)
2023-08-08 21:39:54.952 19169-19169 System.err              com.<packagename>               W  	at com.google.android.material.button.MaterialButton.performClick(MaterialButton.java:1211)
2023-08-08 21:39:54.952 19169-19169 System.err              com.<packagename>               W  	at android.view.View.performClickInternal(View.java:7291)
2023-08-08 21:39:54.952 19169-19169 System.err              com.<packagename>               W  	at android.view.View.access$3600(View.java:838)
2023-08-08 21:39:54.952 19169-19169 System.err              com.<packagename>               W  	at android.view.View$PerformClick.run(View.java:28247)
2023-08-08 21:39:54.952 19169-19169 System.err              com.<packagename>               W  	at android.os.Handler.handleCallback(Handler.java:900)
2023-08-08 21:39:54.952 19169-19169 System.err              com.<packagename>               W  	at android.os.Handler.dispatchMessage(Handler.java:103)
2023-08-08 21:39:54.952 19169-19169 System.err              com.<packagename>               W  	at android.os.Looper.loop(Looper.java:219)
2023-08-08 21:39:54.952 19169-19169 System.err              com.<packagename>               W  	at android.app.ActivityThread.main(ActivityThread.java:8676)
2023-08-08 21:39:54.953 19169-19169 System.err              com.<packagename>               W  	at java.lang.reflect.Method.invoke(Native Method)
2023-08-08 21:39:54.953 19169-19169 System.err              com.<packagename>               W  	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:513)
2023-08-08 21:39:54.953 19169-19169 System.err              com.<packagename>               W  	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1109)`

One last thing I want to note that on desktop Intiface puts out messages into log when a client is connected to it. On mobile - nothing. The last thing it reports is "Listening to 127.0.0.1:12345"
Which is super weird. Because obviously the client does receive a device list, with a correct device being reported with all the correct settings. So there is communication, it's just half-functional

**Comment by AaxAisa** [comment]
Now that I think about it, it feels like my app can read data from websocket but not send messages. It could be my project settings.

**Comment by blackspherefollower** [comment]
Huh, this shouldn't differ between desktop and mobile. Have you enabled "Use foreground process" in Intiface?

Device scanning isn't automatic, you must call StartScanning, but if the server isn't actually handling messages, that might not be processed.

Overall though, I think this is progress and I'll try and promote 3.1 to release tomorrow.

**Comment by AaxAisa** [comment]
So I've done some digging.
In file: ButtplugClientWSClient.java
In method: protected CompletableFuture<ButtplugMessage> sendMessage(final ButtplugMessage msg)
Line: session.getRemote().sendStringByFuture(getParser().formatJson(msg)).get();

Returns an error. If I dig inside, it all falls apart in org/eclipse/jetty/util/FutureCallback.java
in method public Void get() throws InterruptedException, ExecutionException
where _cause is java.nio.channels.ClosedChannelException

Which is a bit weird for me to see, because AFAIK this exception can only be caused if 
a) server closed the connection
b) the client (me) closed a connection, but still tries to use it.

I'm pretty sure neither is the case here.

I can also confirm that the tutorial client (web) can connect to Intiface and causes the logs to post messages, sucessfully finds and connects the device. Can't say if it can control the device - mine is not a vibrator, so :)

**Comment by AaxAisa** [comment]
Upon reading a bit more, it seems like it is possible for the connection to
be open for reading but closed for writing. Which seems like the case for
me. I will try to investigate it in the evening

**Comment by AaxAisa** [comment]
> Huh, this shouldn't differ between desktop and mobile. Have you enabled
"Use foreground process" in Intiface?

I should note that I had not tried your refactor on desktop yet. I tried it
yesterday and ran into weird build problems. It was getting way too late so
I left debugging it for today.
It was a NoClassFound exception btw, sadly can't give you details rn.

As for intiface - yes, everything is enabled and it works with web app.

I will also try to change timeouts for the connection as well as try
connectBlocking instead of connect. Those are the only solutions I can find
online so far.

**Comment by blackspherefollower** [comment]
> Can't say if it can control the device - mine is not a vibrator, so :)

Have you tried https://playground.buttplug.world ? The tutorial is soooo out of date I'm amazed it's even still conencting

**Comment by AaxAisa** [comment]
This one connects no problem too (on my phone). And intiface posts logs.
I'm at work, so no viable deviced (although it could be hilarious to find
one :D.

But my buttplug4j app doesn't produce any logs in intiface, can read the
websocket connection data but cant send it.

**Comment by blackspherefollower** [comment]
Odd... the library passed my tests locally (on desktop), so I'm kinda surprised it's having so much trouble on mobile.
Any chance you can put your app code up so I can debug it?

**Comment by AaxAisa** [comment]
I could do that sure.
It's basically one of your test examples modified and plugged into android
activity. You press a button and it causes your client test to run. That
connects to websocket and then controls the toy.

I didn't change anything elsewhere.

**Comment by AaxAisa** [comment]
Do you have a patch of your refactoring? It currently downloads as a
library, but I'd love to just apply it over current project, overwriting
the files.

**Comment by blackspherefollower** [comment]
Nope, 3.1.101 is what's at master: other than commenting out the disabled annotation to run the client tests, you should be testing with the same library code as me.

**Comment by AaxAisa** [comment]
Ok. It'd just be a bit easier if you printed stack traces when getting
errors. Instead you generate your own error and attach the original to it,
which is OK but only acessible in debug mode. I wanted to change that, but
I can absolutely live with it :D

**Comment by AaxAisa** [comment]
I fixed it, and it was a very silly and a bit obscure error.
Basically, since 3.0 Android is very strict with its UI thread. And that includes prohibiting network connections. So when it detects there is actually a network connection attempt in the main (UI) thread - it freaks out and and throws an exception. That in turn causes all network connections to close.

I'm not blaming here or anything - my code was very basic and silly. I just wanted on button press to connect, move device, disconnect - we're done. I didn't know that I couldn't do that. And unfortunately as I said before your code obscures the error messages and doesn't print stack trace on exception. But it does "Handle" that exception, so it does not prolifirate to my client class where I can catch and study it.

TLDR I found it by making A LOT of breakpoints. But yeah. Your code works.

I would still suggest printing stack traces of actual exception that was thrown, rather than wrap it into your own exception class that only prints "Unknown exception" or "WS was bad"
Again, I'm not blaming or attacking, just noting.

Cheers! Your library is still a literal life saver! :D

**Comment by blackspherefollower** [comment]
Ok then. Let's call this fixed by 850e0ce294432475d633cbcff8be804798f388bb

I'll open a follow-on issue for doing something better with exceptions (they should just be chained, so you can get to the whole implementation specific exception, but either I got that wrong or I need sample logic to demonstrate it)
